### PR TITLE
Backport RPN stack removal from #2178

### PR DESCRIPTION
**Pull request by jtramm**
This PR is a backport of changes by @myerspat that removed the RPN vector that was used as a stack when evaluating complex CSG cells.

This was found to have a major impact on both performance and practicality of some fusion problems, where complex CSG regions are often used, and CSG regions can commonly be bounded by hundreds of surfaces.

This PR is part of a series of several PRs that I'm putting together to support fixed source transport for fusion problems.

### CODE REVIEW

File: src/cell.cpp [hunk 1]
```diff
@@ -110,14 +110,100 @@ tokenize(const std::string region_spec) {
...
+  // Add left parenthesis
+  if (start_token == OP_INTERSECTION) {
+    start = infix.insert(start - 1, OP_LEFT_PAREN);
+  } else {
+    start = infix.insert(start + 1, OP_LEFT_PAREN);
+  }
+  start++;
+
+  // Initialize return iterator
+  auto return_iterator = infix.begin();
```
Comment (paulromano) [inline src/cell.cpp]: There were some further changes in here from openmc-dev/openmc#2318 that should be included.

File: src/cell.cpp [hunk 2]
```diff
@@ -1020,7 +1124,7 @@ UniversePartitioner::UniversePartitioner(const Universe& univ)
       }
     }
 
-    // If there are no bounding z-planes, add this cell to all partitions.
+    // If thregion are no bounding z-planes, add this cell to all partitions.
```
Comment (paulromano) [inline src/cell.cpp]:
```suggestion
    // If there are no bounding z-planes, add this cell to all partitions.
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Looks good -- thanks for backporting this! As noted below, there was a bugfix that hit the `develop` branch on a different PR that should be included too.

**Comment by jtramm** [comment]
Thanks for letting me know about the extra patches on this -- I'll add them in!

File: src/cell.cpp (hunk 1 above)
Comment (jtramm) [reply src/cell.cpp]: I've added in the changes. There's a few very minor differences owing to the OpenMP port not having the `Region` abstraction yet, but is generally 1:1.

File: src/cell.cpp [hunk 3]
```diff
@@ -706,31 +797,34 @@ Cell::to_hdf5(hid_t cell_group) const
 
 BoundingBox Cell::bounding_box_simple() const {
   BoundingBox bbox;
-  for (int32_t token : rpn_) {
-    //bbox &= model::surfaces[abs(token)-1].bounding_box(token > 0);
-    bbox &= model::surfaces[std::abs(token)-1].bounding_box(token > 0);
+  for (int32_t token : region_) {
+    bbox &= model::surfaces[abs(token) - 1].bounding_box(token > 0);
```
Comment (paulromano) [inline src/cell.cpp:801]: What was the deal with `abs` vs `std::abs` here?

**Review by paulromano — APPROVED** [review]
Thanks for the updates! I'll go ahead and merge, although curious about the one question below.

File: src/cell.cpp (hunk 3 above)
Comment (jtramm) [reply src/cell.cpp:801]: Good question! It looks like at some point fairly early in the port I thought it was necessary to switch everything to `std::abs()` (see dcd603a). My best guess is that this was a workaround for a compiler bug. Still -- seems worthwhile to switch it back to std::abs for consistency perhaps? The change to `abs()` in this PR to this branch was not intentionally done, it was just picking up what was there from the cherry pick of the original PR's merge commit.

**Pull request merged**